File: rust/tests/test_rle.rs

extern crate ipaddress;

#[cfg(test)]
mod tests {
    use ipaddress::rle;
    use ipaddress::rle::Rle;
    #[test]
    pub fn test_rle() {
        let empty : Vec<u16> = Vec::new();
        assert_eq!(rle::code(&empty), []);
        assert_eq!(rle::code(&vec![4711]), [Rle {
                        part: 4711,
                        pos: 0,
                        cnt: 1,
                        max: true,
                    }]);
        assert_eq!(rle::code(&vec![4711, 4711]), [Rle {
                        part: 4711,
                        pos: 0,
                        cnt: 2,
                        max: true,
                    }]);
        assert_eq!(rle::code(&vec![4711, 4711, 4811]), [Rle {
                        part: 4711,
                        pos: 0,
                        cnt: 2,
                        max: true,
                    },
                    Rle {
                        part: 4811,
                        pos: 1,
                        cnt: 1,
                        max: true,
                    }]);
        assert_eq!(rle::code(&vec![4711, 4711, 4811, 4711, 4711]), [Rle {
                        part: 4711,
                        pos: 0,
                        cnt: 2,
                        max: true,
                    },
                    Rle {
                        part: 4811,
                        pos: 1,
                        cnt: 1,
                        max: true,
                    },
                    Rle {
                        part: 4711,
                        pos: 2,
                        cnt: 2,
                        max: true,
                    }]);
        assert_eq!(rle::code(&vec![4711, 4711, 4711, 4811, 4711, 4711]), [Rle {
                        part: 4711,
                        pos: 0,
                        cnt: 3,
                        max: true,
                    },
                    Rle {
                        part: 4811,
                        pos: 1,
                        cnt: 1,
                        max: true,
                    },
                    Rle {
                        part: 4711,
                        pos: 2,
                        cnt: 2,
                        max: false,
                    }]
                   );
           assert_eq!(rle::code(&vec![4711, 4711, 4711, 4811, 4711, 4711, 4911, 4911, 4911]), [Rle {
                           part: 4711,
                           pos: 0,
                           cnt: 3,
                           max: true,
                       },
                       Rle {
                           part: 4811,
                           pos: 1,
                           cnt: 1,
                           max: true,
                       },
                       Rle {
                           part: 4711,
                           pos: 2,
                           cnt: 2,
                           max: false,
                       },
                       Rle {
                           part: 4911,
                           pos: 3,
                           cnt: 3,
                           max: true,
                       }]);


       assert_eq!(rle::code(&vec![0x2001, 0x888, 0, 0x6630, 0, 0, 0, 0]), [Rle {
                       part: 0x2001,
                       pos: 0,
                       cnt: 1,
                       max: true,
                   },
                   Rle {
                       part: 0x888,
                       pos: 1,
                       cnt: 1,
                       max: true,
                   },
                   Rle {
                       part: 0,
                       pos: 2,
                       cnt: 1,
                       max: false,
                   },
                   Rle {
                       part: 0x6630,
                       pos: 3,
                       cnt: 1,
                       max: true,
                   },
                   Rle {
                       part: 0,
                       pos: 4,
                       cnt: 4,
                       max: true,
                   }
       ]);


    }
}

File: python/tests/test_rle.py
from ip_address.rle import Rle


def assert_rle(left, right):
    assert len(left) == len(right), "array length missmatch"
    for i in range(len(left)):
        assert left[i].eq(right[i]), f"left:{left[i]} right:{right[i]}"


def test_rle():
    assert_rle(Rle.code([]), [])
    assert_rle(Rle.code([4711]), [Rle(4711, 0, 1, True)])
    assert_rle(Rle.code([4711, 4711]), [Rle(4711, 0, 2, True)])
    assert_rle(
        Rle.code([4711, 4711, 4811]),
        [
            Rle(4711, 0, 2, True),
            Rle(4811, 1, 1, True),
        ],
    )
    assert_rle(
        Rle.code([4711, 4711, 4811, 4711, 4711]),
        [
            Rle(4711, 0, 2, True),
            Rle(4811, 1, 1, True),
            Rle(4711, 2, 2, True),
        ],
    )
    assert_rle(
        Rle.code([4711, 4711, 4711, 4811, 4711, 4711]),
        [
            Rle(4711, 0, 3, True),
            Rle(4811, 1, 1, True),
            Rle(4711, 2, 2, False),
        ],
    )
    assert_rle(
        Rle.code([4711, 4711, 4711, 4811, 4711, 4711, 4911, 4911, 4911]),
        [
            Rle(4711, 0, 3, True),
            Rle(4811, 1, 1, True),
            Rle(4711, 2, 2, False),
            Rle(4911, 3, 3, True),
        ],
    )

    assert_rle(
        Rle.code([0x2001, 0x888, 0, 0x6630, 0, 0, 0, 0]),
        [
            Rle(0x2001, 0, 1, True),
            Rle(0x888, 1, 1, True),
            Rle(0, 2, 1, False),
            Rle(0x6630, 3, 1, True),
            Rle(0, 4, 4, True),
        ],
    )
